[8.0] ADD account_invoice_rounding_by_currency

eLBati opened the pull request:
Unit rounded invoice (Swedish rounding) by currency
===================================================

This module extends functionality of module Unit rounded invoice https://github.com/OCA/account-invoicing/tree/8.0/account_invoice_rounding

It allows to set, in accounting settings, a rounding precision for each currency,
such as 0.05 CHF for Swiss invoices

Configuration
=============

In Settings -> Configuration -> Accounting you will find
Currencies Rounding Rules
Set currency rule for each currency you need to handle.

- `Swedish Round globally`

  To round your invoice total amount, this option will do the adjustment in
  the most important tax line of your invoice.

- `Swedish Round by adding an invoice line`

  To round your invoice total amount, this option create a invoice line without
  taxes on it. This invoice line is tagged as `is_rounding`

archetipo committed:
[IMP] init new module

[FIX] module and description

[FIX] Flake8
[IMP] test case
[IMP] security rule
[IMP] ITA Translation

[FIX] flake8

[FIX] remarks

[FIX] Copyright
[FIX] Configuration
[FIX] translations
[FIX] README
[FIX] remarks

[FIX] remarks

elicoidal commented on account_invoice_rounding_by_currency/README.rst:
```diff
@@ -0,0 +1,73 @@
+   :alt: License: AGPL-3
+
+===================================================
+Unit rounded invoice (Swedish rounding) by currency
```
Is it officially called "Swedish Rounding"?

eLBati replied:
Hi @elicoidal the term is already used here https://github.com/OCA/account-invoicing/tree/8.0/account_invoice_rounding

See https://en.wikipedia.org/wiki/Cash_rounding

elicoidal replied:
I see that but it is NZ English ;)
I would suggest to switch to "Cash Rounding" which is more in line with American English used by Odoo.

elicoidal commented on account_invoice_rounding_by_currency/README.rst:
```diff
@@ -0,0 +1,73 @@
+This module extends functionality of module `Unit rounded invoice <https://github.com/OCA/account-invoicing/tree/8.0/account_invoice_rounding>`_.
+
+It allows to set, in accounting settings, a rounding precision for each currency,
+such as 0.05 CHF for Swiss invoices
```
please add a period

eLBati replied:
Done

elicoidal commented on account_invoice_rounding_by_currency/README.rst:
```diff
@@ -0,0 +1,73 @@
+Configuration
+=============
+
+In Settings -> Configuration -> Accounting you will find
```
Please use enumerated list for procedures

eLBati replied:
done

elicoidal commented on account_invoice_rounding_by_currency/README.rst:
```diff
@@ -0,0 +1,73 @@
+
+- `Swedish Round by adding an invoice line`
+
+  To round your invoice total amount, this option create a invoice line without
```
"this option will create an invoice line without tax on it."
add a period at the end of the sentence

eLBati replied:
I don't understand this one

elicoidal replied:
To round your invoice total amount, this option create a invoice line this option will create an invoice line without tax on it.
This invoice line is tagged as `is_rounding`.
(Period=".")

elicoidal commented on account_invoice_rounding_by_currency/demo/data.xml:
```diff
@@ -0,0 +1,107 @@
+    <data noupdate="1">
+
+        <record id="tax_code_10" model="account.tax.code">
+            <field name="name">IVA a credito 10%</field>
```
please use English in the xml files and add a translation file

eLBati replied:
Done

elicoidal commented on account_invoice_rounding_by_currency/models/company.py:
```diff
@@ -0,0 +1,49 @@
+        help="If you select 'Round per line' : for each tax, the tax "
+        "amount will first be computed and rounded for each "
+        "PO/SO/invoice line and then these rounded amounts will be "
+        "summed, leading to the total amount for that tax. If you "
```
"summed up" (2 times)

eLBati replied:
Doesn't "sum up" mean "give a summary"?

elicoidal replied:
It does but you have the other meaning as well. Check #3 here: http://www.thefreedictionary.com/sum+up
"sum up - determine the sum ofsum up - determine the sum of; "Add all the people in this town to those of the neighboring town""

Anyway, I am not here to be picky and I am not native English so it is up to you :)

elicoidal commented on account_invoice_rounding_by_currency/tests/test_rounding_by_currencies.py:
```diff
@@ -0,0 +1,81 @@
+# -*- coding: utf-8 -*-
```
Please use short headers

eLBati replied:
Done

elicoidal reviewed (changes requested):
Thanks for the contribution.
Mostly small details

eLBati committed:
IMP account_invoice_rounding_by_currency meta data

and FIX
odoo/openerp/fields.py", line 624, in _setup_dependency
fields = [model._fields[head]]
KeyError: 'type'

FIX test: There is no period defined for this date: 2015-10-07

eLBati commented:
@elicoidal thanks for the review. I made changes

eLBati commented:
@elicoidal thanks, made the changes

eLBati commented:
:green_heart: green now

hurrinico commented:
LGTM :+1:

tafaRU commented:
@elicoidal, could you please have a final review before to merge it?
Thank you in advance.

elicoidal commented on account_invoice_rounding_by_currency/README.rst:
```diff
@@ -0,0 +1,81 @@
+#. Accounting
+
+you will find
+Currencies Rounding Rules
```
I would suggest something like: 
```
#. in Settings > Configuration > Accounting, check Currencies Rounding Rules
#. Set currency rule for each currency you need to handle:

  - `Swedish Round globally`
    To round your invoice total amount, this option will do the adjustment in
    the most important tax line of your invoice.
  - `Swedish Round by adding an invoice line`
    To round your invoice total amount, this option will create a invoice line without
    taxes on it.
    This invoice line is tagged as `is_rounding`

```

elicoidal commented on account_invoice_rounding_by_currency/README.rst:
```diff
@@ -0,0 +1,81 @@
+  This invoice line is tagged as `is_rounding`
+
+
+
```
remove extra lines

elicoidal commented on account_invoice_rounding_by_currency/models/company.py:
```diff
@@ -0,0 +1,50 @@
+
+    _sql_constraints = [
+        ('currency_id_uniq_per_company', 'unique (currency_id, company_id)',
+            'Currency must be unique per Company!'),
```
Usually better to avoid the "!" which makes the end user uncomfortable :wink:

elicoidal reviewed (approved):
some unblocking details

eLBati commented:
@elicoidal thanks, done
